File: 2 semester/hw1/1.2/src/com/group144/zaicev/List.java
package com.group144.zaicev;
import java.util.Scanner;

public class List<Type> {

    private class ListElement {
        Type value;
        ListElement next;
        ListElement(Type value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }

    private ListElement head = null;
    private int size = 0;

    public void add(Type value) {
        if (size == 0) {
            this.head = new ListElement(value, null);
            this.size++;
            return;
        }

        ListElement newElement = new ListElement(value, null);
        ListElement tempElement = head;
        while (tempElement.next != null) {
            tempElement = tempElement.next;
        }

        tempElement.next = newElement;
        size++;
    }

    public void printList() {
        if (size == 0) {
            return;
        }

        ListElement current = head;
        while (current.next != null) {
            System.out.print(current.value + " ");
            current = current.next;
        }
        System.out.println(current.value);
    }

    public int size() {
        return size;
    }

    public void pop() {
        if (size > 1) {
            ListElement tempElement = head;
            while (tempElement.next.next != null) {
                tempElement = tempElement.next;
            }
            tempElement.next = null;
        } else if (size == 0) {
            return;
        } else {
            head = null;
        }
        size--;
    }

    public void clear() {
        while (size != 0) {
            pop();
        }
    }

    public Type valueOfIndex(int index) {
        ListElement current = head;
        int i = 0;
        while (i != index) {
            i++;
            current = current.next;
        }
        return current.value;
    }
}

File: first semester/test/test2/3/list.cpp
#include <iostream>
#include "list.h"

struct ListElement
{
    int value;
    ListElement *next;
};

struct List
{
    ListElement *head;
    int size;
};

List *createList()
{
    List *newList = new List;
    newList->head = nullptr;
    newList->size = 0;
    return newList;
}

void addFirstElement(int value, List *list)
{
    ListElement *newElement = new ListElement;
    newElement->next = nullptr;
    newElement->value = value;
    list->head = newElement;
    list->size = 1;
}

void addList(int value, List *list)
{
    if (list->size == 0)
    {
        addFirstElement(value, list);
        return;
    }

    ListElement *newElement = new ListElement;

    newElement->value = value;
    newElement->next = nullptr;
    ListElement *tempElement = list->head;
    while (tempElement->next != nullptr)
        tempElement = tempElement->next;
    tempElement->next = newElement;
    (list->size)++;
}

void printList(List *list)
{
    if (list->size == 0)
        return;

    ListElement *current = list->head;
    while (current->next != nullptr)
    {
        std::cout << current->value << ' ';
        current = current->next;
    }
    std::cout << current->value;
}

int size(List *list)
{
    return list->size;
}

int value(ListElement *element)
{
    return element->value;
}


void pop(List* list)
{
    if (list->size > 1)
    {
        ListElement* temp = list->head;
        while (temp->next->next != nullptr)
        {
            temp = temp->next;
        }
        delete temp->next;
        temp->next = nullptr;
    }
    else if (list->size == 0)
        return;
    else
    {
        delete list->head;
        list->head = nullptr;
    }
    list->size--;
}

void clear(List *list)
{
    while (list->size != 0)
        pop(list);
}

void deleteList(List *list)
{
    clear(list);
    delete list->head;
    delete list;
}

int valueOfIndex(List *list, int index)
{
    ListElement *current = list->head;
    int i = 0;
    while (i != index)
    {
        i++;
        current = current->next;
        //std::cout << "kek \n";
    }
    return current->value;
}

void change(List *list, int value, int index)
{
    ListElement *current = list->head;
    int currentIndex = 0;
    while (currentIndex != index)
    {
        current = current->next;
        currentIndex++;
    }
    current->value = value;
}
